// clicksend/lib/Models/Account.js

/**
 * clicksend
 *
 * This file was automatically generated for ClickSend by APIMATIC v2.0 ( https://apimatic.io ) on 12/20/2016
 */
var BaseModel = require("./BaseModel");
/**
 * Creates a instance of Account
 *
 * @constructor
 */
Account = function (obj) {
    if(!obj) {
        this.username = null;     
        this.password = null;     
        this.userPhone = null;     
        this.userEmail = null;     
        this.userFirstName = null;     
        this.userLastName = null;     
        this.accountName = null;     
        this.country = null;     
        //Append to variable dictionary
        this._variableDict['userPhone'] = 'user_phone';
        this._variableDict['userEmail'] = 'user_email';
        this._variableDict['userFirstName'] = 'user_first_name';
        this._variableDict['userLastName'] = 'user_last_name';
        this._variableDict['accountName'] = 'account_name';
    } else {
        this.username = obj.username;
        this.password = obj.password;
        this.userPhone = obj.user_phone;
        this.userEmail = obj.user_email;
        this.userFirstName = obj.user_first_name;
        this.userLastName = obj.user_last_name;
        this.accountName = obj.account_name;
        this.country = obj.country;
        //Append to variable dictionary
        this._variableDict['userPhone'] = 'user_phone';
        this._variableDict['userEmail'] = 'user_email';
        this._variableDict['userFirstName'] = 'user_first_name';
        this._variableDict['userLastName'] = 'user_last_name';
        this._variableDict['accountName'] = 'account_name';
    }
};

Account.prototype = new BaseModel();
Account.prototype.constructor = Account;

/**
 * Your username
 *
 * @return {string}
 */
Account.prototype.getUsername = function() {
    return this.username;
};

/**
 * Setter for Username
 * 
 * @param {string} value 
 */
Account.prototype.setUsername = function(value) {
    this.username = value;
};

/**
 * Your password
 *
 * @return {string}
 */
Account.prototype.getPassword = function() {
    return this.password;
};

/**
 * Setter for Password
 * 
 * @param {string} value 
 */
Account.prototype.setPassword = function(value) {
    this.password = value;
};

/**
 * Your phone number in E.164 format.
 *
 * @return {string}
 */
Account.prototype.getUserPhone = function() {
    return this.userPhone;
};

/**
 * Setter for UserPhone
 * 
 * @param {string} value 
 */
Account.prototype.setUserPhone = function(value) {
    this.userPhone = value;
};

/**
 * Your email
 *
 * @return {string}
 */
Account.prototype.getUserEmail = function() {
    return this.userEmail;
};

/**
 * Setter for UserEmail
 * 
 * @param {string} value 
 */
Account.prototype.setUserEmail = function(value) {
    this.userEmail = value;
};

/**
 * Your first name
 *
 * @return {string}
 */
Account.prototype.getUserFirstName = function() {
    return this.userFirstName;
};

/**
 * Setter for UserFirstName
 * 
 * @param {string} value 
 */
Account.prototype.setUserFirstName = function(value) {
    this.userFirstName = value;
};

/**
 * Your last name
 *
 * @return {string}
 */
Account.prototype.getUserLastName = function() {
    return this.userLastName;
};

/**
 * Setter for UserLastName
 * 
 * @param {string} value 
 */
Account.prototype.setUserLastName = function(value) {
    this.userLastName = value;
};

/**
 * Your delivery to value.
 *
 * @return {string}
 */
Account.prototype.getAccountName = function() {
    return this.accountName;
};

/**
 * Setter for AccountName
 * 
 * @param {string} value 
 */
Account.prototype.setAccountName = function(value) {
    this.accountName = value;
};

/**
 * Your country
 *
 * @return {string}
 */
Account.prototype.getCountry = function() {
    return this.country;
};

/**
 * Setter for Country
 * 
 * @param {string} value 
 */
Account.prototype.setCountry = function(value) {
    this.country = value;
};


module.exports = Account;
