// clicksend/lib/Models/Address.js

/**
 * clicksend
 *
 * This file was automatically generated for ClickSend by APIMATIC v2.0 ( https://apimatic.io ) on 12/20/2016
 */
var BaseModel = require("./BaseModel");
/**
 * Creates a instance of Address
 *
 * @constructor
 */
Address = function (obj) {
    if(!obj) {
        this.addressName = null;     
        this.addressLine1 = null;     
        this.addressCity = null;     
        this.addressState = null;     
        this.addressPostalCode = null;     
        this.addressCountry = null;     
        this.addressLine2 = null;     
        //Append to variable dictionary
        this._variableDict['addressName'] = 'address_name';
        this._variableDict['addressLine1'] = 'address_line_1';
        this._variableDict['addressCity'] = 'address_city';
        this._variableDict['addressState'] = 'address_state';
        this._variableDict['addressPostalCode'] = 'address_postal_code';
        this._variableDict['addressCountry'] = 'address_country';
        this._variableDict['addressLine2'] = 'address_line_2';
    } else {
        this.addressName = obj.address_name;
        this.addressLine1 = obj.address_line_1;
        this.addressCity = obj.address_city;
        this.addressState = obj.address_state;
        this.addressPostalCode = obj.address_postal_code;
        this.addressCountry = obj.address_country;
        this.addressLine2 = obj.address_line_2;
        //Append to variable dictionary
        this._variableDict['addressName'] = 'address_name';
        this._variableDict['addressLine1'] = 'address_line_1';
        this._variableDict['addressCity'] = 'address_city';
        this._variableDict['addressState'] = 'address_state';
        this._variableDict['addressPostalCode'] = 'address_postal_code';
        this._variableDict['addressCountry'] = 'address_country';
        this._variableDict['addressLine2'] = 'address_line_2';
    }
};

Address.prototype = new BaseModel();
Address.prototype.constructor = Address;

/**
 * Your address name.
 *
 * @return {string}
 */
Address.prototype.getAddressName = function() {
    return this.addressName;
};

/**
 * Setter for AddressName
 * 
 * @param {string} value 
 */
Address.prototype.setAddressName = function(value) {
    this.addressName = value;
};

/**
 * Your address line 1
 *
 * @return {string}
 */
Address.prototype.getAddressLine1 = function() {
    return this.addressLine1;
};

/**
 * Setter for AddressLine1
 * 
 * @param {string} value 
 */
Address.prototype.setAddressLine1 = function(value) {
    this.addressLine1 = value;
};

/**
 * Your city
 *
 * @return {string}
 */
Address.prototype.getAddressCity = function() {
    return this.addressCity;
};

/**
 * Setter for AddressCity
 * 
 * @param {string} value 
 */
Address.prototype.setAddressCity = function(value) {
    this.addressCity = value;
};

/**
 * Your state
 *
 * @return {string}
 */
Address.prototype.getAddressState = function() {
    return this.addressState;
};

/**
 * Setter for AddressState
 * 
 * @param {string} value 
 */
Address.prototype.setAddressState = function(value) {
    this.addressState = value;
};

/**
 * Your postal code
 *
 * @return {int}
 */
Address.prototype.getAddressPostalCode = function() {
    return this.addressPostalCode;
};

/**
 * Setter for AddressPostalCode
 * 
 * @param {int} value 
 */
Address.prototype.setAddressPostalCode = function(value) {
    this.addressPostalCode = value;
};

/**
 * Your country
 *
 * @return {string}
 */
Address.prototype.getAddressCountry = function() {
    return this.addressCountry;
};

/**
 * Setter for AddressCountry
 * 
 * @param {string} value 
 */
Address.prototype.setAddressCountry = function(value) {
    this.addressCountry = value;
};

/**
 * Your address line 2
 *
 * @return {string|null}
 */
Address.prototype.getAddressLine2 = function() {
    return this.addressLine2;
};

/**
 * Setter for AddressLine2
 * 
 * @param {string|null} value 
 */
Address.prototype.setAddressLine2 = function(value) {
    this.addressLine2 = value;
};


module.exports = Address;
