// clicksend/lib/Models/Contact.js

/**
 * clicksend
 *
 * This file was automatically generated for ClickSend by APIMATIC v2.0 ( https://apimatic.io ) on 12/20/2016
 */
var BaseModel = require("./BaseModel");
/**
 * Creates a instance of Contact
 *
 * @constructor
 */
Contact = function (obj) {
    if(!obj) {
        this.phoneNumber = null;     
        this.custom1 = null;     
        this.email = null;     
        this.faxNumber = null;     
        this.firstName = null;     
        this.addressLine1 = null;     
        this.addressLine2 = null;     
        this.addressCity = null;     
        this.addressState = null;     
        this.addressPostalCode = null;     
        this.addressCountry = null;     
        this.organizationName = null;     
        this.custom2 = null;     
        this.custom3 = null;     
        this.custom4 = null;     
        this.lastName = null;     
        //Append to variable dictionary
        this._variableDict['phoneNumber'] = 'phone_number';
        this._variableDict['custom1'] = 'custom_1';
        this._variableDict['faxNumber'] = 'fax_number';
        this._variableDict['firstName'] = 'first_name';
        this._variableDict['addressLine1'] = 'address_line_1';
        this._variableDict['addressLine2'] = 'address_line_2';
        this._variableDict['addressCity'] = 'address_city';
        this._variableDict['addressState'] = 'address_state';
        this._variableDict['addressPostalCode'] = 'address_postal_code';
        this._variableDict['addressCountry'] = 'address_country';
        this._variableDict['organizationName'] = 'organization_name';
        this._variableDict['custom2'] = 'custom_2';
        this._variableDict['custom3'] = 'custom_3';
        this._variableDict['custom4'] = 'custom_4';
        this._variableDict['lastName'] = 'last_name';
    } else {
        this.phoneNumber = obj.phone_number;
        this.custom1 = obj.custom_1;
        this.email = obj.email;
        this.faxNumber = obj.fax_number;
        this.firstName = obj.first_name;
        this.addressLine1 = obj.address_line_1;
        this.addressLine2 = obj.address_line_2;
        this.addressCity = obj.address_city;
        this.addressState = obj.address_state;
        this.addressPostalCode = obj.address_postal_code;
        this.addressCountry = obj.address_country;
        this.organizationName = obj.organization_name;
        this.custom2 = obj.custom_2;
        this.custom3 = obj.custom_3;
        this.custom4 = obj.custom_4;
        this.lastName = obj.last_name;
        //Append to variable dictionary
        this._variableDict['phoneNumber'] = 'phone_number';
        this._variableDict['custom1'] = 'custom_1';
        this._variableDict['faxNumber'] = 'fax_number';
        this._variableDict['firstName'] = 'first_name';
        this._variableDict['addressLine1'] = 'address_line_1';
        this._variableDict['addressLine2'] = 'address_line_2';
        this._variableDict['addressCity'] = 'address_city';
        this._variableDict['addressState'] = 'address_state';
        this._variableDict['addressPostalCode'] = 'address_postal_code';
        this._variableDict['addressCountry'] = 'address_country';
        this._variableDict['organizationName'] = 'organization_name';
        this._variableDict['custom2'] = 'custom_2';
        this._variableDict['custom3'] = 'custom_3';
        this._variableDict['custom4'] = 'custom_4';
        this._variableDict['lastName'] = 'last_name';
    }
};

Contact.prototype = new BaseModel();
Contact.prototype.constructor = Contact;

/**
 * Your phone number in E.164 format. Must be provided if no fax number or email.
 *
 * @return {string}
 */
Contact.prototype.getPhoneNumber = function() {
    return this.phoneNumber;
};

/**
 * Setter for PhoneNumber
 * 
 * @param {string} value 
 */
Contact.prototype.setPhoneNumber = function(value) {
    this.phoneNumber = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string}
 */
Contact.prototype.getCustom1 = function() {
    return this.custom1;
};

/**
 * Setter for Custom1
 * 
 * @param {string} value 
 */
Contact.prototype.setCustom1 = function(value) {
    this.custom1 = value;
};

/**
 * Your email. Must be provided if no phone number or fax number.
 *
 * @return {string|null}
 */
Contact.prototype.getEmail = function() {
    return this.email;
};

/**
 * Setter for Email
 * 
 * @param {string|null} value 
 */
Contact.prototype.setEmail = function(value) {
    this.email = value;
};

/**
 * You fax number. Must be provided if no phone number or email.
 *
 * @return {string|null}
 */
Contact.prototype.getFaxNumber = function() {
    return this.faxNumber;
};

/**
 * Setter for FaxNumber
 * 
 * @param {string|null} value 
 */
Contact.prototype.setFaxNumber = function(value) {
    this.faxNumber = value;
};

/**
 * Your first name.
 *
 * @return {string|null}
 */
Contact.prototype.getFirstName = function() {
    return this.firstName;
};

/**
 * Setter for FirstName
 * 
 * @param {string|null} value 
 */
Contact.prototype.setFirstName = function(value) {
    this.firstName = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string|null}
 */
Contact.prototype.getAddressLine1 = function() {
    return this.addressLine1;
};

/**
 * Setter for AddressLine1
 * 
 * @param {string|null} value 
 */
Contact.prototype.setAddressLine1 = function(value) {
    this.addressLine1 = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string|null}
 */
Contact.prototype.getAddressLine2 = function() {
    return this.addressLine2;
};

/**
 * Setter for AddressLine2
 * 
 * @param {string|null} value 
 */
Contact.prototype.setAddressLine2 = function(value) {
    this.addressLine2 = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string|null}
 */
Contact.prototype.getAddressCity = function() {
    return this.addressCity;
};

/**
 * Setter for AddressCity
 * 
 * @param {string|null} value 
 */
Contact.prototype.setAddressCity = function(value) {
    this.addressCity = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string|null}
 */
Contact.prototype.getAddressState = function() {
    return this.addressState;
};

/**
 * Setter for AddressState
 * 
 * @param {string|null} value 
 */
Contact.prototype.setAddressState = function(value) {
    this.addressState = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string|null}
 */
Contact.prototype.getAddressPostalCode = function() {
    return this.addressPostalCode;
};

/**
 * Setter for AddressPostalCode
 * 
 * @param {string|null} value 
 */
Contact.prototype.setAddressPostalCode = function(value) {
    this.addressPostalCode = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string|null}
 */
Contact.prototype.getAddressCountry = function() {
    return this.addressCountry;
};

/**
 * Setter for AddressCountry
 * 
 * @param {string|null} value 
 */
Contact.prototype.setAddressCountry = function(value) {
    this.addressCountry = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string|null}
 */
Contact.prototype.getOrganizationName = function() {
    return this.organizationName;
};

/**
 * Setter for OrganizationName
 * 
 * @param {string|null} value 
 */
Contact.prototype.setOrganizationName = function(value) {
    this.organizationName = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string|null}
 */
Contact.prototype.getCustom2 = function() {
    return this.custom2;
};

/**
 * Setter for Custom2
 * 
 * @param {string|null} value 
 */
Contact.prototype.setCustom2 = function(value) {
    this.custom2 = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string|null}
 */
Contact.prototype.getCustom3 = function() {
    return this.custom3;
};

/**
 * Setter for Custom3
 * 
 * @param {string|null} value 
 */
Contact.prototype.setCustom3 = function(value) {
    this.custom3 = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string|null}
 */
Contact.prototype.getCustom4 = function() {
    return this.custom4;
};

/**
 * Setter for Custom4
 * 
 * @param {string|null} value 
 */
Contact.prototype.setCustom4 = function(value) {
    this.custom4 = value;
};

/**
 * Your last name
 *
 * @return {string|null}
 */
Contact.prototype.getLastName = function() {
    return this.lastName;
};

/**
 * Setter for LastName
 * 
 * @param {string|null} value 
 */
Contact.prototype.setLastName = function(value) {
    this.lastName = value;
};


module.exports = Contact;
