// clicksend/lib/Models/ContactList.js

/**
 * clicksend
 *
 * This file was automatically generated for ClickSend by APIMATIC v2.0 ( https://apimatic.io ) on 12/20/2016
 */
var BaseModel = require("./BaseModel");
/**
 * Creates a instance of ContactList
 *
 * @constructor
 */
ContactList = function (obj) {
    if(!obj) {
        this.listId = null;     
        this.listName = null;     
        this.listEmailId = null;     
        this.contactsCount = null;     
        //Append to variable dictionary
        this._variableDict['listId'] = 'list_id';
        this._variableDict['listName'] = 'list_name';
        this._variableDict['listEmailId'] = 'list_email_id';
        this._variableDict['contactsCount'] = 'contacts_count';
    } else {
        this.listId = obj.list_id;
        this.listName = obj.list_name;
        this.listEmailId = obj.list_email_id;
        this.contactsCount = obj.contacts_count;
        //Append to variable dictionary
        this._variableDict['listId'] = 'list_id';
        this._variableDict['listName'] = 'list_name';
        this._variableDict['listEmailId'] = 'list_email_id';
        this._variableDict['contactsCount'] = 'contacts_count';
    }
};

ContactList.prototype = new BaseModel();
ContactList.prototype.constructor = ContactList;

/**
 * TODO: Write general description for this method
 *
 * @return {string}
 */
ContactList.prototype.getListId = function() {
    return this.listId;
};

/**
 * Setter for ListId
 * 
 * @param {string} value 
 */
ContactList.prototype.setListId = function(value) {
    this.listId = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string|null}
 */
ContactList.prototype.getListName = function() {
    return this.listName;
};

/**
 * Setter for ListName
 * 
 * @param {string|null} value 
 */
ContactList.prototype.setListName = function(value) {
    this.listName = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string|null}
 */
ContactList.prototype.getListEmailId = function() {
    return this.listEmailId;
};

/**
 * Setter for ListEmailId
 * 
 * @param {string|null} value 
 */
ContactList.prototype.setListEmailId = function(value) {
    this.listEmailId = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {int|null}
 */
ContactList.prototype.getContactsCount = function() {
    return this.contactsCount;
};

/**
 * Setter for ContactsCount
 * 
 * @param {int|null} value 
 */
ContactList.prototype.setContactsCount = function(value) {
    this.contactsCount = value;
};


module.exports = ContactList;
