// clicksend/lib/Models/FaxMessage.js

/**
 * clicksend
 *
 * This file was automatically generated for ClickSend by APIMATIC v2.0 ( https://apimatic.io ) on 12/20/2016
 */
var BaseModel = require("./BaseModel");
/**
 * Creates a instance of FaxMessage
 *
 * @constructor
 */
FaxMessage = function (obj) {
    if(!obj) {
        this.source = null;     
        this.to = null;     
        this.listId = null;     
        this.from = null;     
        this.schedule = null;     
        this.customString = null;     
        this.country = null;     
        this.fromEmail = null;     
        //Append to variable dictionary
        this._variableDict['listId'] = 'list_id';
        this._variableDict['customString'] = 'custom_string';
        this._variableDict['fromEmail'] = 'from_email';
    } else {
        this.source = obj.source;
        this.to = obj.to;
        this.listId = obj.list_id;
        this.from = obj.from;
        this.schedule = obj.schedule;
        this.customString = obj.custom_string;
        this.country = obj.country;
        this.fromEmail = obj.from_email;
        //Append to variable dictionary
        this._variableDict['listId'] = 'list_id';
        this._variableDict['customString'] = 'custom_string';
        this._variableDict['fromEmail'] = 'from_email';
    }
};

FaxMessage.prototype = new BaseModel();
FaxMessage.prototype.constructor = FaxMessage;

/**
 * Your method of sending e.g. 'wordpress', 'php', 'c#'.
 *
 * @return {string}
 */
FaxMessage.prototype.getSource = function() {
    return this.source;
};

/**
 * Setter for Source
 * 
 * @param {string} value 
 */
FaxMessage.prototype.setSource = function(value) {
    this.source = value;
};

/**
 * Recipient fax number in E.164 format.
 *
 * @return {string}
 */
FaxMessage.prototype.getTo = function() {
    return this.to;
};

/**
 * Setter for To
 * 
 * @param {string} value 
 */
FaxMessage.prototype.setTo = function(value) {
    this.to = value;
};

/**
 * Your list ID if sending to a whole list. Can be used instead of 'to'.
 *
 * @return {int|null}
 */
FaxMessage.prototype.getListId = function() {
    return this.listId;
};

/**
 * Setter for ListId
 * 
 * @param {int|null} value 
 */
FaxMessage.prototype.setListId = function(value) {
    this.listId = value;
};

/**
 * Your sender id. Must be a valid fax number.
 *
 * @return {string|null}
 */
FaxMessage.prototype.getFrom = function() {
    return this.from;
};

/**
 * Setter for From
 * 
 * @param {string|null} value 
 */
FaxMessage.prototype.setFrom = function(value) {
    this.from = value;
};

/**
 * Leave blank for immediate delivery. Your schedule time in unix format http://help.clicksend.com/what-is-a-unix-timestamp
 *
 * @return {int|null}
 */
FaxMessage.prototype.getSchedule = function() {
    return this.schedule;
};

/**
 * Setter for Schedule
 * 
 * @param {int|null} value 
 */
FaxMessage.prototype.setSchedule = function(value) {
    this.schedule = value;
};

/**
 * Your reference. Will be passed back with all replies and delivery reports.
 *
 * @return {string|null}
 */
FaxMessage.prototype.getCustomString = function() {
    return this.customString;
};

/**
 * Setter for CustomString
 * 
 * @param {string|null} value 
 */
FaxMessage.prototype.setCustomString = function(value) {
    this.customString = value;
};

/**
 * Recipient country.
 *
 * @return {string|null}
 */
FaxMessage.prototype.getCountry = function() {
    return this.country;
};

/**
 * Setter for Country
 * 
 * @param {string|null} value 
 */
FaxMessage.prototype.setCountry = function(value) {
    this.country = value;
};

/**
 * An email address where the reply should be emailed to.
 *
 * @return {string|null}
 */
FaxMessage.prototype.getFromEmail = function() {
    return this.fromEmail;
};

/**
 * Setter for FromEmail
 * 
 * @param {string|null} value 
 */
FaxMessage.prototype.setFromEmail = function(value) {
    this.fromEmail = value;
};


module.exports = FaxMessage;
