// clicksend/lib/Models/MmsMessage.js

/**
 * clicksend
 *
 * This file was automatically generated for ClickSend by APIMATIC v2.0 ( https://apimatic.io ) on 12/20/2016
 */
var BaseModel = require("./BaseModel");
/**
 * Creates a instance of MmsMessage
 *
 * @constructor
 */
MmsMessage = function (obj) {
    if(!obj) {
        this.messages = null;     
        this.mediaFile = null;     
        //Append to variable dictionary
        this._variableDict['mediaFile'] = 'media_file';
    } else {
        this.messages = obj.messages;
        this.mediaFile = obj.media_file;
        //Append to variable dictionary
        this._variableDict['mediaFile'] = 'media_file';
    }
};

MmsMessage.prototype = new BaseModel();
MmsMessage.prototype.constructor = MmsMessage;

/**
 * TODO: Write general description for this method
 *
 * @return {object}
 */
MmsMessage.prototype.getMessages = function() {
    return this.messages;
};

/**
 * Setter for Messages
 * 
 * @param {object} value 
 */
MmsMessage.prototype.setMessages = function(value) {
    this.messages = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string}
 */
MmsMessage.prototype.getMediaFile = function() {
    return this.mediaFile;
};

/**
 * Setter for MediaFile
 * 
 * @param {string} value 
 */
MmsMessage.prototype.setMediaFile = function(value) {
    this.mediaFile = value;
};


module.exports = MmsMessage;
