// clicksend/lib/Models/ReferralAccount.js

/**
 * clicksend
 *
 * This file was automatically generated for ClickSend by APIMATIC v2.0 ( https://apimatic.io ) on 12/20/2016
 */
var BaseModel = require("./BaseModel");
/**
 * Creates a instance of ReferralAccount
 *
 * @constructor
 */
ReferralAccount = function (obj) {
    if(!obj) {
        this.referralRuleId = null;     
        this.referedUserId = null;     
        this.dateReferred = null;     
        this.percentageReferral = null;     
        //Append to variable dictionary
        this._variableDict['referralRuleId'] = 'referral_rule_id';
        this._variableDict['referedUserId'] = 'refered_user_id';
        this._variableDict['dateReferred'] = 'date_referred';
        this._variableDict['percentageReferral'] = 'percentage_referral';
    } else {
        this.referralRuleId = obj.referral_rule_id;
        this.referedUserId = obj.refered_user_id;
        this.dateReferred = obj.date_referred;
        this.percentageReferral = obj.percentage_referral;
        //Append to variable dictionary
        this._variableDict['referralRuleId'] = 'referral_rule_id';
        this._variableDict['referedUserId'] = 'refered_user_id';
        this._variableDict['dateReferred'] = 'date_referred';
        this._variableDict['percentageReferral'] = 'percentage_referral';
    }
};

ReferralAccount.prototype = new BaseModel();
ReferralAccount.prototype.constructor = ReferralAccount;

/**
 * TODO: Write general description for this method
 *
 * @return {string}
 */
ReferralAccount.prototype.getReferralRuleId = function() {
    return this.referralRuleId;
};

/**
 * Setter for ReferralRuleId
 * 
 * @param {string} value 
 */
ReferralAccount.prototype.setReferralRuleId = function(value) {
    this.referralRuleId = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string}
 */
ReferralAccount.prototype.getReferedUserId = function() {
    return this.referedUserId;
};

/**
 * Setter for ReferedUserId
 * 
 * @param {string} value 
 */
ReferralAccount.prototype.setReferedUserId = function(value) {
    this.referedUserId = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string}
 */
ReferralAccount.prototype.getDateReferred = function() {
    return this.dateReferred;
};

/**
 * Setter for DateReferred
 * 
 * @param {string} value 
 */
ReferralAccount.prototype.setDateReferred = function(value) {
    this.dateReferred = value;
};

/**
 * TODO: Write general description for this method
 *
 * @return {string}
 */
ReferralAccount.prototype.getPercentageReferral = function() {
    return this.percentageReferral;
};

/**
 * Setter for PercentageReferral
 * 
 * @param {string} value 
 */
ReferralAccount.prototype.setPercentageReferral = function(value) {
    this.percentageReferral = value;
};


module.exports = ReferralAccount;
