// clicksend/lib/Models/SmsCampaign.js

/**
 * clicksend
 *
 * This file was automatically generated for ClickSend by APIMATIC v2.0 ( https://apimatic.io ) on 12/20/2016
 */
var BaseModel = require("./BaseModel");
/**
 * Creates a instance of SmsCampaign
 *
 * @constructor
 */
SmsCampaign = function (obj) {
    if(!obj) {
        this.listId = null;     
        this.name = null;     
        this.body = null;     
        this.schedule = null;     
        this.from = null;     
        //Append to variable dictionary
        this._variableDict['listId'] = 'list_id';
    } else {
        this.listId = obj.list_id;
        this.name = obj.name;
        this.body = obj.body;
        this.schedule = obj.schedule;
        this.from = obj.from;
        //Append to variable dictionary
        this._variableDict['listId'] = 'list_id';
    }
};

SmsCampaign.prototype = new BaseModel();
SmsCampaign.prototype.constructor = SmsCampaign;

/**
 * Your list id.
 *
 * @return {object}
 */
SmsCampaign.prototype.getListId = function() {
    return this.listId;
};

/**
 * Setter for ListId
 * 
 * @param {object} value 
 */
SmsCampaign.prototype.setListId = function(value) {
    this.listId = value;
};

/**
 * Your campaign name.
 *
 * @return {string}
 */
SmsCampaign.prototype.getName = function() {
    return this.name;
};

/**
 * Setter for Name
 * 
 * @param {string} value 
 */
SmsCampaign.prototype.setName = function(value) {
    this.name = value;
};

/**
 * Your campaign message.
 *
 * @return {string}
 */
SmsCampaign.prototype.getBody = function() {
    return this.body;
};

/**
 * Setter for Body
 * 
 * @param {string} value 
 */
SmsCampaign.prototype.setBody = function(value) {
    this.body = value;
};

/**
 * Your schedule timestamp.
 *
 * @return {string}
 */
SmsCampaign.prototype.getSchedule = function() {
    return this.schedule;
};

/**
 * Setter for Schedule
 * 
 * @param {string} value 
 */
SmsCampaign.prototype.setSchedule = function(value) {
    this.schedule = value;
};

/**
 * Your sender id - more info: http://help.clicksend.com/SMS/what-is-a-sender-id-or-sender-number.
 *
 * @return {string|null}
 */
SmsCampaign.prototype.getFrom = function() {
    return this.from;
};

/**
 * Setter for From
 * 
 * @param {string|null} value 
 */
SmsCampaign.prototype.setFrom = function(value) {
    this.from = value;
};


module.exports = SmsCampaign;
