// clicksend/lib/Models/SmsMessage.js

/**
 * clicksend
 *
 * This file was automatically generated for ClickSend by APIMATIC v2.0 ( https://apimatic.io ) on 12/20/2016
 */
var BaseModel = require("./BaseModel");
/**
 * Creates a instance of SmsMessage
 *
 * @constructor
 */
SmsMessage = function (obj) {
    if(!obj) {
        this.source = null;     
        this.from = null;     
        this.body = null;     
        this.to = null;     
        this.schedule = null;     
        this.customString = null;     
        this.listId = null;     
        this.country = null;     
        this.fromEmail = null;     
        //Append to variable dictionary
        this._variableDict['customString'] = 'custom_string';
        this._variableDict['listId'] = 'list_id';
        this._variableDict['fromEmail'] = 'from_email';
    } else {
        this.source = obj.source;
        this.from = obj.from;
        this.body = obj.body;
        this.to = obj.to;
        this.schedule = obj.schedule;
        this.customString = obj.custom_string;
        this.listId = obj.list_id;
        this.country = obj.country;
        this.fromEmail = obj.from_email;
        //Append to variable dictionary
        this._variableDict['customString'] = 'custom_string';
        this._variableDict['listId'] = 'list_id';
        this._variableDict['fromEmail'] = 'from_email';
    }
};

SmsMessage.prototype = new BaseModel();
SmsMessage.prototype.constructor = SmsMessage;

/**
 * Your method of sending e.g. 'wordpress', 'php', 'c#'.
 *
 * @return {string}
 */
SmsMessage.prototype.getSource = function() {
    return this.source;
};

/**
 * Setter for Source
 * 
 * @param {string} value 
 */
SmsMessage.prototype.setSource = function(value) {
    this.source = value;
};

/**
 * Your sender id - more info: http://help.clicksend.com/SMS/what-is-a-sender-id-or-sender-number.
 *
 * @return {string}
 */
SmsMessage.prototype.getFrom = function() {
    return this.from;
};

/**
 * Setter for From
 * 
 * @param {string} value 
 */
SmsMessage.prototype.setFrom = function(value) {
    this.from = value;
};

/**
 * Your message.
 *
 * @return {string}
 */
SmsMessage.prototype.getBody = function() {
    return this.body;
};

/**
 * Setter for Body
 * 
 * @param {string} value 
 */
SmsMessage.prototype.setBody = function(value) {
    this.body = value;
};

/**
 * Recipient phone number in E.164 format.
 *
 * @return {string}
 */
SmsMessage.prototype.getTo = function() {
    return this.to;
};

/**
 * Setter for To
 * 
 * @param {string} value 
 */
SmsMessage.prototype.setTo = function(value) {
    this.to = value;
};

/**
 * Leave blank for immediate delivery. Your schedule time in unix format http://help.clicksend.com/what-is-a-unix-timestamp
 *
 * @return {int|null}
 */
SmsMessage.prototype.getSchedule = function() {
    return this.schedule;
};

/**
 * Setter for Schedule
 * 
 * @param {int|null} value 
 */
SmsMessage.prototype.setSchedule = function(value) {
    this.schedule = value;
};

/**
 * Your reference. Will be passed back with all replies and delivery reports.
 *
 * @return {string|null}
 */
SmsMessage.prototype.getCustomString = function() {
    return this.customString;
};

/**
 * Setter for CustomString
 * 
 * @param {string|null} value 
 */
SmsMessage.prototype.setCustomString = function(value) {
    this.customString = value;
};

/**
 * Your list ID if sending to a whole list. Can be used instead of 'to'.
 *
 * @return {int|null}
 */
SmsMessage.prototype.getListId = function() {
    return this.listId;
};

/**
 * Setter for ListId
 * 
 * @param {int|null} value 
 */
SmsMessage.prototype.setListId = function(value) {
    this.listId = value;
};

/**
 * Recipient country.
 *
 * @return {string|null}
 */
SmsMessage.prototype.getCountry = function() {
    return this.country;
};

/**
 * Setter for Country
 * 
 * @param {string|null} value 
 */
SmsMessage.prototype.setCountry = function(value) {
    this.country = value;
};

/**
 * An email address where the reply should be emailed to. If omitted, the reply will be emailed back to the user who sent the outgoing SMS.
 *
 * @return {string|null}
 */
SmsMessage.prototype.getFromEmail = function() {
    return this.fromEmail;
};

/**
 * Setter for FromEmail
 * 
 * @param {string|null} value 
 */
SmsMessage.prototype.setFromEmail = function(value) {
    this.fromEmail = value;
};


module.exports = SmsMessage;
