// clicksend/lib/Models/Subaccount.js

/**
 * clicksend
 *
 * This file was automatically generated for ClickSend by APIMATIC v2.0 ( https://apimatic.io ) on 12/20/2016
 */
var BaseModel = require("./BaseModel");
/**
 * Creates a instance of Subaccount
 *
 * @constructor
 */
Subaccount = function (obj) {
    if(!obj) {
        this.apiUsername = null;     
        this.password = null;     
        this.email = null;     
        this.phoneNumber = null;     
        this.firstName = null;     
        this.lastName = null;     
        this.accessUsers  = true;     
        this.accessBilling  = true;     
        this.accessReporting  = true;     
        this.accessContacts  = false;     
        this.accessSettings  = true;     
        //Append to variable dictionary
        this._variableDict['apiUsername'] = 'api_username';
        this._variableDict['phoneNumber'] = 'phone_number';
        this._variableDict['firstName'] = 'first_name';
        this._variableDict['lastName'] = 'last_name';
        this._variableDict['accessUsers'] = 'access_users';
        this._variableDict['accessBilling'] = 'access_billing';
        this._variableDict['accessReporting'] = 'access_reporting';
        this._variableDict['accessContacts'] = 'access_contacts';
        this._variableDict['accessSettings'] = 'access_settings';
    } else {
        this.apiUsername = obj.api_username;
        this.password = obj.password;
        this.email = obj.email;
        this.phoneNumber = obj.phone_number;
        this.firstName = obj.first_name;
        this.lastName = obj.last_name;
        this.accessUsers = obj.access_users;
        this.accessBilling = obj.access_billing;
        this.accessReporting = obj.access_reporting;
        this.accessContacts = obj.access_contacts;
        this.accessSettings = obj.access_settings;
        //Append to variable dictionary
        this._variableDict['apiUsername'] = 'api_username';
        this._variableDict['phoneNumber'] = 'phone_number';
        this._variableDict['firstName'] = 'first_name';
        this._variableDict['lastName'] = 'last_name';
        this._variableDict['accessUsers'] = 'access_users';
        this._variableDict['accessBilling'] = 'access_billing';
        this._variableDict['accessReporting'] = 'access_reporting';
        this._variableDict['accessContacts'] = 'access_contacts';
        this._variableDict['accessSettings'] = 'access_settings';
    }
};

Subaccount.prototype = new BaseModel();
Subaccount.prototype.constructor = Subaccount;

/**
 * Your new api username.
 *
 * @return {array}
 */
Subaccount.prototype.getApiUsername = function() {
    return this.apiUsername;
};

/**
 * Setter for ApiUsername
 * 
 * @param {array} value 
 */
Subaccount.prototype.setApiUsername = function(value) {
    this.apiUsername = value;
};

/**
 * Your new password
 *
 * @return {string}
 */
Subaccount.prototype.getPassword = function() {
    return this.password;
};

/**
 * Setter for Password
 * 
 * @param {string} value 
 */
Subaccount.prototype.setPassword = function(value) {
    this.password = value;
};

/**
 * Your new email.
 *
 * @return {string}
 */
Subaccount.prototype.getEmail = function() {
    return this.email;
};

/**
 * Setter for Email
 * 
 * @param {string} value 
 */
Subaccount.prototype.setEmail = function(value) {
    this.email = value;
};

/**
 * Your phone number in E.164 format.
 *
 * @return {string}
 */
Subaccount.prototype.getPhoneNumber = function() {
    return this.phoneNumber;
};

/**
 * Setter for PhoneNumber
 * 
 * @param {string} value 
 */
Subaccount.prototype.setPhoneNumber = function(value) {
    this.phoneNumber = value;
};

/**
 * Your firstname
 *
 * @return {string}
 */
Subaccount.prototype.getFirstName = function() {
    return this.firstName;
};

/**
 * Setter for FirstName
 * 
 * @param {string} value 
 */
Subaccount.prototype.setFirstName = function(value) {
    this.firstName = value;
};

/**
 * Your lastname
 *
 * @return {string}
 */
Subaccount.prototype.getLastName = function() {
    return this.lastName;
};

/**
 * Setter for LastName
 * 
 * @param {string} value 
 */
Subaccount.prototype.setLastName = function(value) {
    this.lastName = value;
};

/**
 * Your access users flag value, must be 1 or 0.
 *
 * @return {bool|null}
 */
Subaccount.prototype.getAccessUsers = function() {
    return this.accessUsers;
};

/**
 * Setter for AccessUsers
 * 
 * @param {bool|null} value 
 */
Subaccount.prototype.setAccessUsers = function(value) {
    this.accessUsers = value;
};

/**
 * Your access billing flag value, must be 1 or 0.
 *
 * @return {bool|null}
 */
Subaccount.prototype.getAccessBilling = function() {
    return this.accessBilling;
};

/**
 * Setter for AccessBilling
 * 
 * @param {bool|null} value 
 */
Subaccount.prototype.setAccessBilling = function(value) {
    this.accessBilling = value;
};

/**
 * Your access reporting flag value, must be 1 or 0.
 *
 * @return {bool|null}
 */
Subaccount.prototype.getAccessReporting = function() {
    return this.accessReporting;
};

/**
 * Setter for AccessReporting
 * 
 * @param {bool|null} value 
 */
Subaccount.prototype.setAccessReporting = function(value) {
    this.accessReporting = value;
};

/**
 * Your access contacts flag value, must be 1 or 0.
 *
 * @return {bool|null}
 */
Subaccount.prototype.getAccessContacts = function() {
    return this.accessContacts;
};

/**
 * Setter for AccessContacts
 * 
 * @param {bool|null} value 
 */
Subaccount.prototype.setAccessContacts = function(value) {
    this.accessContacts = value;
};

/**
 * Your access settings flag value, must be 1 or 0.
 *
 * @return {bool|null}
 */
Subaccount.prototype.getAccessSettings = function() {
    return this.accessSettings;
};

/**
 * Setter for AccessSettings
 * 
 * @param {bool|null} value 
 */
Subaccount.prototype.setAccessSettings = function(value) {
    this.accessSettings = value;
};


module.exports = Subaccount;
