// clicksend/lib/Models/VoiceMessage.js

/**
 * clicksend
 *
 * This file was automatically generated for ClickSend by APIMATIC v2.0 ( https://apimatic.io ) on 12/20/2016
 */
var BaseModel = require("./BaseModel");
/**
 * Creates a instance of VoiceMessage
 *
 * @constructor
 */
VoiceMessage = function (obj) {
    if(!obj) {
        this.to = null;     
        this.body = null;     
        this.voice = null;     
        this.customString = null;     
        this.country = null;     
        this.source = null;     
        this.listId = null;     
        this.lang = null;     
        this.schedule = null;     
        //Append to variable dictionary
        this._variableDict['customString'] = 'custom_string';
        this._variableDict['listId'] = 'list_id';
    } else {
        this.to = obj.to;
        this.body = obj.body;
        this.voice = obj.voice;
        this.customString = obj.custom_string;
        this.country = obj.country;
        this.source = obj.source;
        this.listId = obj.list_id;
        this.lang = obj.lang;
        this.schedule = obj.schedule;
        //Append to variable dictionary
        this._variableDict['customString'] = 'custom_string';
        this._variableDict['listId'] = 'list_id';
    }
};

VoiceMessage.prototype = new BaseModel();
VoiceMessage.prototype.constructor = VoiceMessage;

/**
 * Your phone number in E.164 format.
 *
 * @return {string}
 */
VoiceMessage.prototype.getTo = function() {
    return this.to;
};

/**
 * Setter for To
 * 
 * @param {string} value 
 */
VoiceMessage.prototype.setTo = function(value) {
    this.to = value;
};

/**
 * Biscuit uv3nlCOjRk croissant chocolate lollipop chocolate muffin.
 *
 * @return {string}
 */
VoiceMessage.prototype.getBody = function() {
    return this.body;
};

/**
 * Setter for Body
 * 
 * @param {string} value 
 */
VoiceMessage.prototype.setBody = function(value) {
    this.body = value;
};

/**
 * Either 'female' or 'male'.
 *
 * @return {string}
 */
VoiceMessage.prototype.getVoice = function() {
    return this.voice;
};

/**
 * Setter for Voice
 * 
 * @param {string} value 
 */
VoiceMessage.prototype.setVoice = function(value) {
    this.voice = value;
};

/**
 * Your reference. Will be passed back with all replies and delivery reports.
 *
 * @return {string}
 */
VoiceMessage.prototype.getCustomString = function() {
    return this.customString;
};

/**
 * Setter for CustomString
 * 
 * @param {string} value 
 */
VoiceMessage.prototype.setCustomString = function(value) {
    this.customString = value;
};

/**
 * The country of the recipient.
 *
 * @return {string}
 */
VoiceMessage.prototype.getCountry = function() {
    return this.country;
};

/**
 * Setter for Country
 * 
 * @param {string} value 
 */
VoiceMessage.prototype.setCountry = function(value) {
    this.country = value;
};

/**
 * Your method of sending e.g. 'wordpress', 'php', 'c#'.
 *
 * @return {string|null}
 */
VoiceMessage.prototype.getSource = function() {
    return this.source;
};

/**
 * Setter for Source
 * 
 * @param {string|null} value 
 */
VoiceMessage.prototype.setSource = function(value) {
    this.source = value;
};

/**
 * Your list ID if sending to a whole list. Can be used instead of 'to'.
 *
 * @return {int|null}
 */
VoiceMessage.prototype.getListId = function() {
    return this.listId;
};

/**
 * Setter for ListId
 * 
 * @param {int|null} value 
 */
VoiceMessage.prototype.setListId = function(value) {
    this.listId = value;
};

/**
 * au (string, required) - See section on available languages.
 *
 * @return {string|null}
 */
VoiceMessage.prototype.getLang = function() {
    return this.lang;
};

/**
 * Setter for Lang
 * 
 * @param {string|null} value 
 */
VoiceMessage.prototype.setLang = function(value) {
    this.lang = value;
};

/**
 * Leave blank for immediate delivery. Your schedule time in unix format http://help.clicksend.com/what-is-a-unix-timestamp
 *
 * @return {int|null}
 */
VoiceMessage.prototype.getSchedule = function() {
    return this.schedule;
};

/**
 * Setter for Schedule
 * 
 * @param {int|null} value 
 */
VoiceMessage.prototype.setSchedule = function(value) {
    this.schedule = value;
};


module.exports = VoiceMessage;
